src/combinatorics/derangements/basic.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.option
+import dynamics.fixed_points.basic
+import group_theory.perm.option
+
+/-!
+# Derangements on types
+
+In this file we define `derangements α`, the set of derangements on a type `α`.
+
+We also define some equivalences involving various subtypes of `perm α` and `derangements α`:
+* `derangements_option_equiv_sigma_at_most_one_fixed_point`: An equivalence between
+  `derangements (option α)` and the sigma-type `Σ a : α, {f : perm α // fixed_points f ⊆ a}`.
+* `derangements_recursion_equiv`: An equivalence between `derangements (option α)` and the
+  sigma-type `Σ a : α, (derangements (({a}ᶜ : set α) : Type _) ⊕ derangements α)` which is later
+  used to inductively count the number of derangements.
+
+In order to prove the above, we also prove some results about the effect of `equiv.remove_none`
+on derangements: `remove_none.fiber_none` and `remove_none.fiber_some`.
+-/
+
+open equiv function
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | ∀ x : α, f x ≠ x}
+
+variables {α β : Type*}
+
+lemma mem_derangements_iff_fixed_points_eq_empty {f : perm α} :
+  f ∈ derangements α ↔ fixed_points f = ∅ :=
+set.eq_empty_iff_forall_not_mem.symm
```

Comment:
You could also add a lemma `mem_derangements_iff_support_eq_top` (or something like that)

Reply:
`function.support` is about `f x ≠ 0` and `function.mul_support` about `f x ≠ 1`. So that doesn't work. However I did think about having `derangements_on (s : set α) := {f : perm α | fixed_points f = sᶜ}` in which case your lemma suggestion could become `derangements α = derangements_on univ`.

Reply:
I thought about the `support` for permutations (on mobile so idk the exact file, but it was used in defining the sign of a permutation), but it was hard to use both that and `function.fixed_points`, so I picked just one. Perhaps the former could be revised to make use of the latter? Outside the scope of this PR though.

Reply:
Well first I see that `equiv.perm.support` supposes `fintype α`, so it's enough for what you're doing here but underpowered otherwise.